### PR TITLE
feat: add dump_stats to advanced profiler (#19698)

### LINKED ISSUE #19698: Dump prof files from AdvancedProfiler
### Description & Motivation

Currently `AdvancedProfiler` only outputs textual summary where the rows are sorted by cumulative time. Unfortunately this does not give a full picture about the structure and duration of call stacks involved.

### Pitch

Raw cProfiler dumps provide more information to help identify bottlenecks than just module names and line numbers. E.g. using https://jiffyclub.github.io/snakeviz one can see the hierarchy of the call stacks and sort/drill-down using an interactive interface.

Luckily there's not much needed to make this available. `AdvancedProfiler` already collects profiling data we can easily dump using [dump_stats()](https://docs.python.org/3/library/profile.html#profile.Profile.dump_stats):

```python
    @override
    def summary(self) -> str:
        recorded_stats = {}
        for action_name, pr in self.profiled_actions.items():
		 	####### dumping prof file ########
            if self.filename and self.dirpath:
                import os
                import tempfile
                from fsspec.core import url_to_fs
                from fsspec.implementations.local import AbstractFileSystem

                filepath = os.path.join(self.dirpath, self._prepare_filename())
                def get_filesystem(path, **kwargs) -> AbstractFileSystem:
                    fs, _ = url_to_fs(str(path), **kwargs)
                    return fs
                with tempfile.TemporaryDirectory(prefix="test", suffix="test", dir=os.getcwd()) as tmp_dir:
                    # alas dump_stats() expects a file path it can open()
                    tmp_file = os.path.join(tmp_dir, "tmp.prof")
                    pr.dump_stats(tmp_file)
                    fs = get_filesystem(filepath)
                    fs.mkdirs(self.dirpath, exist_ok=True)
                    fs.copy(tmp_file, filepath + ".prof")
			###############################
            s = io.StringIO()
            ps = pstats.Stats(pr, stream=s).strip_dirs().sort_stats("cumulative")
            ps.print_stats(self.line_count_restriction)
            recorded_stats[action_name] = s.getvalue()
        return self._stats_to_str(recorded_stats)
```

I'm willing to contribute this but would like to clarify the best configuration and injection options. Thanks!

### Alternatives

Write a custom callback and do the profiling from there - thus not using the one provided by PyTorch Lightning.

### Additional context

Related to https://github.com/Lightning-AI/pytorch-lightning/issues/7424

cc @borda @carmocca

### PR DESCRIPTION
**Pull request by clumsy**
## What does this PR do?

Adding a new `dump_stats` flag to `AdvanedProfiler` to persist `.prof` files collected during profiling. These files can be visualized by [SnakeViz](https://jiffyclub.github.io/snakeviz/) like so:

Screenshot 2024-03-26 at 10 04 17 AM

A `.prof` file is created for each profiled action, e.g.:
```
...
fit-run-[Strategy]DDPStrategy.on_train_end.prof
fit-run-[Callback]TQDMProgressBar.on_fit_end.prof
fit-run-[Callback]ModelSummary.on_fit_end.prof
fit-run-[Callback]ModelCheckpoint{'monitor': None, 'mode': 'min', 'every_n_train_steps': 0, 'every_n_epochs': 1, 'train_time_interval': None}.on_fit_end.prof
fit-run-[Callback]LearningRateMonitor.on_fit_end.prof
...
```

Fixes #19698

No breaking changes, adding a new backward-compatible flag, with option disabled by default.

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by clumsy** [comment]
Just bumping this one in case it got lost in notifications, @awaelchli.

File: src/lightning/pytorch/profilers/advanced.py [hunk 1]
```diff
@@ -54,13 +58,17 @@ def __init__(
...
+
         Raises:
             ValueError:
                 If you attempt to stop recording an action which was never started.
         """
         super().__init__(dirpath=dirpath, filename=filename)
         self.profiled_actions: Dict[str, cProfile.Profile] = {}
         self.line_count_restriction = line_count_restriction
+        self.dump_stats = dump_stats
+        assert not self.dump_stats or self.dirpath is not None, "dirname must be provided for dump_states to work"
```
Comment (awaelchli) [inline src/lightning/pytorch/profilers/advanced.py:71]: Could we make this a ValueError rather than an assertion error? That's because the error would be caused by user input.

**Review by awaelchli — COMMENTED** [review]
@clumsy Looks great to me. Just a few minor comments

File: src/lightning/pytorch/profilers/advanced.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/pytorch/profilers/advanced.py:71]: According to the docs:

```
If ``dirpath`` is ``None`` but ``filename`` is present, the
                ``trainer.log_dir`` (from :class:`~lightning.pytorch.loggers.tensorboard.TensorBoardLogger`)
                will be used.
```
So a None check here looks like is too early. We'd have to move it somewhere later after initialization.

File: src/lightning/pytorch/profilers/advanced.py [hunk 2]
```diff
@@ -75,10 +83,28 @@ def stop(self, action_name: str) -> None:
...
 
+    def _maybe_dump_stats(self, action_name: str, pr: cProfile.Profile) -> None:
+        if not self.dump_stats:
+            return
+        assert self.dirpath  # redundant, but needed for mypy
+        dst_filepath = os.path.join(self.dirpath, self._prepare_filename(action_name=action_name, extension=".prof"))
+        dst_fs = get_filesystem(dst_filepath)
+        dst_fs.mkdirs(self.dirpath, exist_ok=True)
+        # temporarily save to local since pstats can only dump into a local file
+        with tempfile.TemporaryDirectory(prefix="test", suffix="test", dir=os.getcwd()) as tmp_dir, dst_fs.open(
```
[the remark this answers is not in the document]
Comment (clumsy) [reply src/lightning/pytorch/profilers/advanced.py:94]: Not really. The temp directory will be unique for each rank so there will be no collision in the full file path. The directory is removed right after this contextmanager exits. The file name already adds the rank in `_prepare_filename`

File: src/lightning/pytorch/profilers/advanced.py (hunk 1 above)
Comment (clumsy) [reply src/lightning/pytorch/profilers/advanced.py:71]: Ok, so `self.dirpath` is always available then.

**Comment by clumsy** [comment]
Thanks for checking, @awaelchli! I've addressed your comments. Please let me know if the are remaining concerns.

File: src/lightning/pytorch/profilers/advanced.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/pytorch/profilers/advanced.py:91]: For the filename that goes into that directory, yes. But the temp directory itself

`tempfile.TemporaryDirectory(prefix="test", suffix="test", dir=os.getcwd())`

is just a randomized name saved in the current working directory. There is no guarantee that the name of this directory will be different on each rank.

**Comment by awaelchli** [comment]
@clumsy Are you still interested in finishing the PR? I think it's very close to land.

**Comment by clumsy** [comment]
Sure, addressed the comments, @awaelchli

**Comment by awaelchli** [comment]
Nice @clumsy thank you for addressing the remaining comments.
